### ISSUE TITLE
Administration UI

### ISSUE THREAD

**Issue by osma** [NatLibFi/Annif#24]
We could have a UI for administering most/all aspects of an Annif installation. It would most likely be a single-page dynamic web app that uses the REST API to perform admin functions.

**Comment by superciccio** [comment]
Hi, 
I came here from a thread on reddit a guy told me about the administration UI.
I'm still fairly new to python I only have written simple scripts ( connect to firebase download images and then compress them ) for personal use.

I also have a background as Java Web developer, but I want to make a shift to Python so that is that reason.
I've worked mostly with Angular but study by my self also Vue.

My question is, do you need help on this task ? Is something that I can pick up ?

Let me know, I would be happy to help

**Comment by osma** [comment]
Thanks for your kind offer @superciccio !

The current UI is very basic. It was recently reworked a bit by @kinow (the guy from reddit I guess!), changing from jQuery to Vue.JS (see #245). That was merged, but there are still parts of jQuery being used. It would be great to be able to get rid of the rest of jQuery too. I think that's mainly used for performing a couple of different REST API calls. See [this comment](https://github.com/NatLibFi/Annif/pull/245#issuecomment-458504479) for some ideas about replacements. So I'd be happy to receive a PR which removes the jQuery dependency and switches over the functionality that VueJS cannot perform to some more lightweight alternative.

Perhaps more related to this issue at hand, eventually I think we want to extend the UI into more administrative functions. Right now it's just a simple form for testing Annif on different documents. However, many UI admin functionalities will require also the REST API to be extended, as right now it's pretty much a read-only API - the real administration happens by editing `projects.cfg` and executing CLI commands. 

An easy starting point would be to add UI functionality for looking at project configuration, similar to the `list-projects` and `show-project` CLI commands. This should already be possible with the current REST API, because you can use GET on a project to ask for information about it.

**Comment by superciccio** [comment]
Hi @osma 

I have some question

Can I update Vue to latest version 2.6.10. Also I see that eveything is inside home.html, can I also do some refactor and export it in a new file ?

> Perhaps more related to this issue at hand, eventually I think we want to extend the UI into more administrative functions
 Do you have something in mind? Like having pages behind authentication or define user/role in config file ?

> It would be great to be able to get rid of the rest of jQuery too
I'll look into, no problem.

> An easy starting point would be to add UI functionality for looking at project configuration, similar to the list-projects and show-project CLI commands. 
Great !

**Comment by osma** [comment]
> Can I update Vue to latest version 2.6.10

I think this should be fine - does @kinow have a comment?

> Also I see that eveything is inside home.html, can I also do some refactor and export it in a new file ?

Sure, this would be welcome!

**Comment by osma** [comment]
Ah sorry, I missed some of your questions that looked like quotes.

> Do you have something in mind? Like having pages behind authentication or define user/role in config file ?

I think authentication, users and roles are out of scope for now (feel free to disagree, anyone!). I see Annif as mainly a microservice that is used by other systems, not (usually) directly by end users. Any authentication or other types of access control can be handled by a suitable reverse proxy such as nginx. (However, there is limited support for basic access control already: projects can be marked as hidden or private, so they don't show up (hidden) or cannot be used at all (private) via the REST API)

More likely starting points would be adding equivalents of the CLI commands `loadvoc` and `train`, i.e. for loading vocabularies and for training models. Both of these would require extending the REST API.

**Comment by kinow** [comment]
>I think this should be fine - does @kinow have a comment?

+1 from me !

**Comment by kinow** [comment]
>I think authentication, users and roles are out of scope for now (feel free to disagree, anyone!). I see Annif as mainly a microservice that is used by other systems, not (usually) directly by end users. Any authentication or other types of access control can be handled by a suitable reverse proxy such as nginx. (However, there is limited support for basic access control already: projects can be marked as hidden or private, so they don't show up (hidden) or cannot be used at all (private) via the REST API)

I think it's a good approach for now. It can be added later anyway (and makes the web dev part a bit easier too)

**Comment by juhoinkinen** [comment]
For a reference: @mjsuhonos [posted in Annif-users](https://groups.google.com/g/annif-users/c/MNyKXNiuuR8/m/FiurSg3xAAAJ) about his work to create: Cannif: https://github.com/mjsuhonos/Annif-corpora/tree/master/cannif

Currently it implements a UI with Streamlit like this:

<img width="1618" height="808" alt="Image" src="https://github.com/user-attachments/assets/d9f5ca46-02da-46a4-813c-41e9b9f8c8aa" />

There is nice UI component for creating new projects:

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/6e325655-d75d-4a05-ba01-86a88e217cfb" />


However the project-creation UI does not yet function; Annif REST API does not have a method for creating or training projects (https://github.com/NatLibFi/Annif/issues/634).

**Comment by mjsuhonos** [comment]
This discussion overlaps somewhat with #261 -- so I suppose now is a good time to discuss the relationship between Annif and cannif and how they will proceed.

The Annif team has put a lot of hard work into documentation and tutorials around how to use Annif, and cannif is **absolutely not** intended to replace any of that -- it is still essential to understand at an underlying level how projects, vocabs, etc. all work together.

Like many tools, cannif was created to help with my own project management issues, particularly in aggregating evaluation outputs with some graphs and managing ensembles.  It was really inspired by [elasticearch-head](https://mobz.github.io/elasticsearch-head/).  So the challenge of cannif is to be helpful within the constraints of Streamlit and the Annif REST API, without being overly hand-holding and trying to abstract away all complexity within a UI.

The current limitations of the REST API and the challenges in project-creation workflow are a good example: presently, Annif does not identify a vocab (and thus cannot `load-vocab`) unless the vocab_id is defined in a project.  So a vocab does not "exist" without a project, but a project cannot be correctly defined without a vocab.  This is why the UI is tricky to get right, and frankly has ballooned the cannif code a bit (pending an upcoming refactor).

@osma says in #261, and I agree:

> I'm a bit torn on this. I've never thought of the Annif web UI as a separate project, more of an administrative interface for testing models … I think the UI should come along when you install the tool.

From my view, I would rather not place any additional burden on the Annif team to develop and maintain a more complex 'administrative' UI -- I would rather you focus on the interesting stuff that requires knowledge of the internal API and the underlying Python libraries.

I am happy to move cannif to its own repository and maintain it in parallel with Annif, in a way to provide use cases for eg. expanding  the REST API functionality #916, etc.  I am actually also OK with running cli commands from cannif  with `subprocess`, as `eval` [is done presently](https://github.com/mjsuhonos/Annif-corpora/blob/master/cannif/cannif_streamlit.py#L400-L403), same with [`load-vocab`](https://github.com/mjsuhonos/Annif-corpora/blob/master/cannif/cannif_streamlit.py#L427-L429).  I can see a use case here for eg. running training workloads remotely on a VPS, for example.

I guess in summary, I see the responsibility for the REST API staying with the Annif team, and UI for administration tasks remaining with cannif.  It's a single file anyway; you can always just copy it to your Annif directory and run it from there, which is probably what most people will do.
